libraries/chain/finality_core.cpp

```diff
@@ -0,0 +1,338 @@
+   return links.front().target_block_num;
+}
+
+/**
+ *  @pre this->links.empty() == false
+ *  @post none
+ *  @returns latest qc_claim made by the core
+ */
+qc_claim finality_core::latest_qc_claim() const
+{
+   assert(!links.empty()); // Satisfied by invariant 1.
+
+   return qc_claim{.block_num = links.back().target_block_num, .is_strong_qc = links.back().is_link_strong};
+}
+
+/**
+ *  @pre last_final_block_num() <= block_num < current_block_num()
+ *
+ *  @post returned block_ref has block_num() == block_num
+ */
+const block_ref& finality_core::get_block_reference(block_num_type block_num) const
+{
+   assert(last_final_block_num() <= block_num); // Satisfied by precondition.
+   assert(block_num < current_block_num()); // Satisfied by precondition.
+
+   // If refs.empty() == true, then by invariant 3, current_block_num() == last_final_block_num(), 
+   // and therefore it is impossible to satisfy the precondition. So going forward, it is safe to assume refs.empty() == false.
+
+   const size_t ref_index = block_num - last_final_block_num();
+
+   // By the precondition, 0 <= ref_index < (current_block_num() - last_final_block_num()).
+   // Then, by invariant 8, 0 <= ref_index < refs.size().
+
+   assert(ref_index < refs.size()); // Satisfied by justification above.
+
+   return refs[ref_index];
+   // By invariants 4 and 6, tail[ref_index].block_num() == block_num, which satisfies the post-condition.
+}
+
+/**
+ *  @pre links.front().source_block_num <= block_num <= current_block_num()
+ *
+ *  @post returned qc_link has source_block_num == block_num
+ */
+const qc_link& finality_core::get_qc_link_from(block_num_type block_num) const
+{
+   assert(!links.empty()); // Satisfied by invariant 1.
+
+   assert(links.front().source_block_num <= block_num); // Satisfied by precondition.
+   assert(block_num <= current_block_num()); // Satisfied by precondition.
+
+   const size_t link_index = block_num - links.front().source_block_num;
+
+   // By the precondition, 0 <= link_index <= (current_block_num() - links.front().source_block_num).
+   // Then, by invariant 9, 0 <= link_index <= links.size() - 1
+
+   assert(link_index < links.size()); // Satisfied by justification above.
+
+   return links[link_index];
+   // By invariants 7, links[link_index].source_block_num == block_num, which satisfies the post-condition.
+}
+
+/**
+ *  @pre c.latest_qc_claim().block_num <= most_recent_ancestor_with_qc.block_num <= c.current_block_num()
+ * 
+ *  @post std::get<0>(returned_value) <= std::get<1>(returned_value) <= std::get<2>(returned_value) <= most_recent_ancestor_with_qc.block_num
+ *  @post c.last_final_block_num() <= std::get<0>(returned_value)
+ *  @post c.links.front().source_block_num <= std::get<1>(returned_value)
+ *  @post c.final_on_strong_qc_block_num <= std::get<2>(returned_value)
+ */
+std::tuple<block_num_type, block_num_type, block_num_type> get_new_block_numbers(const finality_core& c, const qc_claim& most_recent_ancestor_with_qc)
```

Comment:
Let's make `get_new_block_numbers` a member function of `finality_core`, rather than a free function taking a `const finality_core&`.

Reply:
OK. maybe later!